man/man1/pmpython.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2016 Ken McDonell.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PMPYTHON 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmpython\f1 \- run a python script using a preferred python variant
.SH SYNOPSIS
\f3pmpython\f1
[\f2python-args-and-script\f1]
.SH DESCRIPTION
.B pmpython
provides a way to run python scripts using a customisable python
interpreter, rather than embedding the name of
a particular version of python into each script.
.PP
This can be useful as it allows version-independent python code to be
run anywhere.
All python modules shipped with PCP support versions 2.6 and later (in
the python2 series), and 3.3 and later (in the python3 release series).
.PP
Due to python monitoring and collecting scripts being relatively simple
in PCP (not requiring new modules, language features, etc), it has been
possible to ensure they work for all of the above python versions.
.PP
However, the name of the python interpreter is not always the same,
thus,
it is common for PCP python scripts to use a \(lqshebang\(rq line that
launches the
.B python
interpreter indirectly
as follows:
.PP
.in 1.0i
.ft CR
.nf
#!/usr/bin/env pmpython
.fi
.ft
.in
.PP
.BR env (1)
is used to find the correct path for the
.B pmpython
executable from the user's
.BR $PATH .
.PP
By default the name of the python interpreter is found from the
the value of
.B $PCP_PYTHON_PROG
from the environment (if set) else
from
.IR /etc/pcp.conf .
The latter is the
more typical case where
this value is based on
some heuristics about the platform at the time the PCP packages were
build and
favour the use of
.I python3
in all recent releases of PCP (for those platforms that support it).
.PP
This allows an appropriate name to be used for the python
interpreter instead of a hard-coded
python version name, while still allowing the user to override the python
interpreter as follows:
.PP
.in 1.0i
.ft CR
.nf
$ PCP_PYTHON_PROG=python3 pmpython --version
Python 3.4.2
$ PCP_PYTHON_PROG=python2 pmpython --version
Python 2.7.9
.fi
.ft R
.in
.PP
This is convenient for shipping identical scripts on multiple platforms,
and for testing different python versions with the one script (e.g. in
the case where multiple versions of python are installed, PCP_PYTHON_PROG
can be set in the local environment to override the global setting).
.PP
.B pmpython
is a replacement for an earlier tool with similar function, namely
.BR pcp-python (1).
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR env (1),
.BR pcp (1),
.BR pcp-python (1)
and
.BR PCPIntro (1).

.\" control lines for scripts/man-spell
.\" +ok+ args


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmpython/pmpython.c
/*
 * Wrapper to execute the version of python(1) defined in the
 * pcp.conf file.
 *
 * Copyright (c) 2016 Ken McDonell.  All Rights Reserved.
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include "pmapi.h"

int
main(int argc, char **argv)
{
    char	*python;
    char	*myname;

    python = pmGetConfig("PCP_PYTHON_PROG");
    if (python[0] == '\0') {
	char *altconf = getenv("PCP_CONF");
	char *altdir = getenv("PCP_DIR");
	if (altconf) {
	    if (altdir)
		fprintf(stderr, "%s: PCP_PYTHON_PROG not set in %s/%s\n", argv[0], altdir, altconf);
	    else
		fprintf(stderr, "%s: PCP_PYTHON_PROG not set in %s\n", argv[0], altconf);
	}
	else {
	    if (altdir)
		fprintf(stderr, "%s: PCP_PYTHON_PROG not set in %s/etc/pcp.conf\n", argv[0], altdir);
	    else
		fprintf(stderr, "%s: PCP_PYTHON_PROG not set in /etc/pcp.conf\n", argv[0]);
	}
	return 1;
    }
    myname = argv[0];
    argv[0] = python;
    execvp(argv[0], argv);

    /* should not get here */
    fprintf(stderr, "%s: execvp(%s, ...) failed: %s\n", myname, python, strerror(errno));

    return 1;
}
